File: esj-core/src/main/java/de/bsnsoft/esj/validate/Finding.java
package de.bsnsoft.esj.validate;

import de.bsnsoft.esj.SemanticPath;
import java.util.Objects;

/**
 * One thing a validator has to say about a document (specification, section 9.5).
 *
 * <p>A caller acts on a finding by its {@link #code()} and by the place it names, so the
 * place is fixed by the specification and not by this implementation. {@link #path()} is
 * that place wherever a semantic path can be it. {@link #subject()} carries what the path
 * cannot: the member name as the document writes it where the defect is about a name that
 * is no path, and the identifier of the term or group where the defect is about one that
 * is missing from the instance the path names or occurs too often in it. Four mandatory
 * terms missing at the root of a document draw four findings with one code and one path,
 * and only the subject tells them apart without the English message being read.
 *
 * @param path     the semantic path the finding is about; the root path, whose text is
 *                 the empty string, stands for a finding about the document as a whole
 * @param subject  what the finding is about where the path cannot name it, or the empty
 *                 string where the path names it
 * @param code     the stable identifier of the kind of problem
 * @param severity how much the finding weighs
 * @param message  human-readable English text
 */
public record Finding(SemanticPath path,
                      String subject,
                      FindingCode code,
                      Severity severity,
                      String message) {

    /**
     * Checks that every part is present.
     *
     * @param path     the semantic path the finding is about; the root path, whose text is
     *                 the empty string, stands for a finding about the document as a whole
     * @param subject  what the finding is about where the path cannot name it, or the empty
     *                 string where the path names it
     * @param code     the stable identifier of the kind of problem
     * @param severity how much the finding weighs
     * @param message  human-readable English text
     * @throws NullPointerException if a part is {@code null}
     */
    public Finding {
        Objects.requireNonNull(path, "path");
        Objects.requireNonNull(subject, "subject");
        Objects.requireNonNull(code, "code");
        Objects.requireNonNull(severity, "severity");
        Objects.requireNonNull(message, "message");
    }

    /**
     * Creates a finding whose path names the place, with the severity the specification
     * fixes for the code.
     *
     * @param path    the semantic path the finding is about
     * @param code    the kind of problem
     * @param message human-readable English text
     * @return the finding, with an empty subject
     */
    public static Finding of(SemanticPath path, FindingCode code, String message) {
        return about(path, "", code, message);
    }

    /**
     * Creates a finding that names what it is about beside the path, with the severity
     * the specification fixes for the code.
     *
     * @param path    the semantic path the finding is about
     * @param subject the member name or the term or group identifier the finding is
     *                about, as the specification, section 9.5 fixes it for the code
     * @param code    the kind of problem
     * @param message human-readable English text
     * @return the finding
     */
    public static Finding about(SemanticPath path,
                                String subject,
                                FindingCode code,
                                String message) {
        return new Finding(path, subject, code, code.defaultSeverity(), message);
    }

    /**
     * Creates a finding about the document as a whole.
     *
     * @param code    the kind of problem
     * @param message human-readable English text
     * @return the finding, carrying the root path and an empty subject
     */
    public static Finding ofDocument(FindingCode code, String message) {
        return of(SemanticPath.root(), code, message);
    }

    /**
     * Tells whether this finding makes the document fail its layer.
     *
     * @return {@code true} if the severity is {@link Severity#ERROR}
     */
    public boolean isError() {
        return severity == Severity.ERROR;
    }

    /**
     * Returns the finding as one line: the code, the path and the message. The subject
     * is not repeated here, because the message that goes with a subject names it.
     *
     * @return a short description of the finding
     */
    @Override
    public String toString() {
        return code.code() + " [" + severity.token() + "] " + path + ": " + message;
    }
}

File: bindings/typescript/src/finding.ts
import type { FindingCode, Layer, NotEvaluatedReason, Severity, ValidationStatus } from './codes.ts';
import { NOT_EVALUATED_CODES, layerOf, severityOf } from './codes.ts';
import { escapeForMessage } from './grammars.ts';

/**
 * A finding, and the result that carries a set of them (specification, section 9.5).
 *
 * A caller acts on `code` and on the place the finding names, and the message is for a
 * person: it may quote the document, and a fragment quoted in it is escaped, so that a
 * message is safe to write to a log line as it stands.
 */

/** One finding. */
export interface Finding {
  /** The semantic path the finding is about, or the empty string. */
  readonly path: string;
  /** What the finding is about where `path` cannot name it, or the empty string. */
  readonly subject: string;
  /** The kind of problem (section 9.6). */
  readonly code: FindingCode;
  /** How much the finding weighs at the layer it belongs to. */
  readonly severity: Severity;
  /** English text, with any quoted fragment escaped. */
  readonly message: string;
}

/** Builds a finding, filling in the severity the specification gives the code. */
export function finding(
  code: FindingCode, message: string, place?: { path?: string; subject?: string },
): Finding {
  return {
    path: place?.path ?? '',
    subject: place?.subject ?? '',
    code,
    severity: severityOf(code),
    message,
  };
}

/** Returns the member access a finding names a member of `values` by (section 9.5). */
export function valuesSubject(name: string): string {
  return 'values["' + escapeForMessage(name) + '"]';
}

/** Returns the member access a finding names a member of `extensions` by (section 9.5). */
export function extensionsSubject(name: string): string {
  return 'extensions["' + escapeForMessage(name) + '"]';
}

/** What a run says about a layer it did not evaluate. */
export interface NotEvaluated {
  /** The layer. */
  readonly layer: Layer;
  /** Why it was not evaluated; the vocabulary is closed (section 9.5). */
  readonly reason: NotEvaluatedReason;
}

/** What a validator returns (section 9.5). */
export interface ValidationResult {
  /** The status, decided by the rules of section 9.5. */
  readonly status: ValidationStatus;
  /** The findings, in the order the validator produced them. */
  readonly findings: readonly Finding[];
  /** The layers that were evaluated. */
  readonly evaluated: readonly Layer[];
  /** The layers that were not, each with the reason. */
  readonly notEvaluated: readonly NotEvaluated[];
}

/**
 * Decides the status of a result from its findings and from what it evaluated
 * (section 9.5).
 *
 * The status is not a function of the severities alone: `ESJ-L1-LIMIT` is an error and
 * yields `INDETERMINATE`, because it says that this implementation stopped and not that the
 * document is wrong, and a result that names a layer as not evaluated is never `VALID`
 * however quiet it is.
 *
 * @param findings the findings
 * @param notEvaluated the layers that were not evaluated
 * @return the status
 */
export function statusOf(
  findings: readonly Finding[], notEvaluated: readonly NotEvaluated[],
): ValidationStatus {
  for (const entry of findings) {
    if (entry.severity === 'error' && !NOT_EVALUATED_CODES.includes(entry.code)) {
      return 'INVALID';
    }
  }
  if (notEvaluated.length > 0) {
    return 'INDETERMINATE';
  }
  for (const entry of findings) {
    if (NOT_EVALUATED_CODES.includes(entry.code)) {
      return 'INDETERMINATE';
    }
  }
  return 'VALID';
}

/** Returns the findings of one layer. */
export function findingsOf(findings: readonly Finding[], layer: Layer): Finding[] {
  return findings.filter((entry) => layerOf(entry.code) === layer);
}
